mergify_engine/merge_train.py

```diff
@@ -0,0 +1,440 @@
+
+    class Serialized(typing.TypedDict):
+        pull_request_number: github_types.GitHubPullRequestNumber
+        queue_name: rules.QueueName
+        parents: typing.List[github_types.GitHubPullRequestNumber]
+        initial_current_base_sha: github_types.SHAType
+        current_base_sha: github_types.SHAType
+        tmp_pull_number: typing.Optional[github_types.GitHubPullRequestNumber]
+
+    def serialized(self) -> "TrainCar.Serialized":
+        return self.Serialized(
+            pull_request_number=self.pull_request_number,
+            queue_name=self.queue_name,
+            parents=self.parents,
+            initial_current_base_sha=self.initial_current_base_sha,
+            current_base_sha=self.current_base_sha,
+            tmp_pull_number=self.tmp_pull_number,
+        )
+
+    @classmethod
+    def deserialized(cls, train: "Train", data: "TrainCar.Serialized") -> "TrainCar":
+        return cls(train, **data)
+
+    def create_pull(self) -> None:
+        # TODO(sileht): reuse branch instead of recreating PRs ?
+        branch_name = f"{constants.MERGE_QUEUE_BRANCH_PREFIX}/{self.train.ref}/{self.pull_request_number}"
+
+        self.train.client.post(
+            f"/repos/{self.train.owner}/{self.train.repo}/git/refs",
+            json={
+                "ref": f"refs/heads/{branch_name}",
+                "sha": self.initial_current_base_sha,
+            },
+        )
+
+        for pull_number in self.parents + [self.pull_request_number]:
+            # TODO(sileht): if a merge fail we should update the summary of self.pull_request_number with
+            # the failure
+            self.train.client.post(
+                f"/repos/{self.train.owner}/{self.train.repo}/merges",
+                json={
+                    "base": branch_name,
+                    "head": f"refs/pull/{pull_number}/head",
+                    "commit_message": f"Merge of #{pull_number}",
+                },
+            )
+
+        try:
+            pulls_for_title = ", ".join(
+                [f"#{p}" for p in self.parents + [self.pull_request_number]]
+            )
+            title = f"merge-queue: testing {self.initial_current_base_sha[:7]}, {pulls_for_title} together"
+            body = ""
+            tmp_pull = self.train.client.post(
+                f"/repos/{self.train.owner}/{self.train.repo}/pulls",
+                json={
+                    "title": title,
+                    "body": body,
+                    "base": self.train.ref,
+                    "head": branch_name,
+                },
+            ).json()
+        except http.HTTPClientSideError as e:
+            if e.status_code == 422 and "No commits between" in e.message:
+                # FIXME(sileht): Delete created branch
+                return
+            raise
+
+        self.tmp_pull_number = tmp_pull["number"]
+
+        sub = utils.async_run(
+            subscription.Subscription.get_subscription(self.train.owner_id)
+        )
```

Comment:
Don't we have a context in the caller? 🤔 
It feels weird to create a context here but maybe.

Reply:
We create a context just to use `check_api.py`, we may need to create a cleaner `check_api.py` to be able to use it with and without a context.